pkg/operator/observe_config.go

```diff
@@ -0,0 +1,175 @@
+package operator
+
+import (
+	"fmt"
+	"reflect"
+	"time"
+
+	"k8s.io/apimachinery/pkg/runtime"
+
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+
+	"github.com/golang/glog"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	kubeinformers "k8s.io/client-go/informers"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/flowcontrol"
+	"k8s.io/client-go/util/workqueue"
+
+	operatorconfigclientv1alpha1 "github.com/openshift/cluster-openshift-apiserver-operator/pkg/generated/clientset/versioned/typed/openshiftapiserver/v1alpha1"
+	operatorconfiginformerv1alpha1 "github.com/openshift/cluster-openshift-apiserver-operator/pkg/generated/informers/externalversions/openshiftapiserver/v1alpha1"
+)
+
+type ConfigObserver struct {
+	operatorConfigClient operatorconfigclientv1alpha1.OpenshiftapiserverV1alpha1Interface
+
+	kubeClient kubernetes.Interface
+
+	// queue only ever has one item, but it has nice error handling backoff/retry semantics
+	queue workqueue.RateLimitingInterface
+
+	rateLimiter flowcontrol.RateLimiter
+}
+
+func NewConfigObserver(
+	operatorConfigInformer operatorconfiginformerv1alpha1.OpenShiftAPIServerOperatorConfigInformer,
+	kubeAPIServerNamespacedKubeInformers kubeinformers.SharedInformerFactory,
+	operatorConfigClient operatorconfigclientv1alpha1.OpenshiftapiserverV1alpha1Interface,
+	kubeClient kubernetes.Interface,
+) *ConfigObserver {
+	c := &ConfigObserver{
+		operatorConfigClient: operatorConfigClient,
+		kubeClient:           kubeClient,
+
+		queue: workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "ConfigObserver"),
+
+		rateLimiter: flowcontrol.NewTokenBucketRateLimiter(0.05 /*3 per minute*/, 4),
+	}
+
+	operatorConfigInformer.Informer().AddEventHandler(c.eventHandler())
+	kubeAPIServerNamespacedKubeInformers.Core().V1().ConfigMaps().Informer().AddEventHandler(c.eventHandler())
+	kubeAPIServerNamespacedKubeInformers.Core().V1().ServiceAccounts().Informer().AddEventHandler(c.eventHandler())
+
+	return c
+}
+
+func (c ConfigObserver) sync() error {
+	operatorConfig, err := c.operatorConfigClient.OpenShiftAPIServerOperatorConfigs().Get("instance", metav1.GetOptions{})
+	if err != nil {
+		return err
+	}
+	kubeAPIServerPublicInfo, err := c.kubeClient.CoreV1().ConfigMaps(kubeAPIServerNamespaceName).Get("public-info", metav1.GetOptions{})
+	if err != nil {
+		return err
+	}
+	observedConfig := map[string]interface{}{}
+
+	if val, ok := kubeAPIServerPublicInfo.Data["projectConfig.defaultNodeSelector"]; ok {
```

Comment:
Mind adding a word of comment why we always stomp `defaultNodeSelector`?

Reply:
> Mind adding a word of comment why we always stomp `defaultNodeSelector`?

sure.